modules/sdk-lib-mpc/src/tss/ecdsa-dkls/dsg.ts

```diff
@@ -0,0 +1,165 @@
+    if (!this.dsgSession) {
+      this.dsgSession = SignSession.fromBytes(this.dsgSessionBytes);
+    }
+  }
+
+  private _deserializeState() {
+    if (!this.dsgSession) {
+      throw Error('Session not intialized');
+    }
+    const round = decode(this.dsgSession.toBytes()).round;
+    switch (round) {
+      case 'WaitMsg1':
+        this.dsgState = DsgState.Round1;
+        break;
+      case 'WaitMsg2':
+        this.dsgState = DsgState.Round2;
+        break;
+      case 'WaitMsg3':
+        this.dsgState = DsgState.Round3;
+        break;
+      case 'Ended':
+        this.dsgState = DsgState.Complete;
+        break;
+      default:
+        this.dsgState = DsgState.InvalidState;
+        throw `Invalid State: ${round}`;
+    }
+  }
+
+  async initDsg(): Promise<DeserializedBroadcastMessage> {
+    if (this.dsgState !== DsgState.Uninitialized) {
+      throw 'DSG session already initialized';
+    }
+    this.dsgSession = new SignSession(this.keyShare, this.derivationPath);
+    try {
+      const payload = this.dsgSession.createFirstMessage().payload;
+      this._deserializeState();
+      return {
+        payload: payload,
+        from: this.partyIdx,
+      };
+    } catch (e) {
+      throw `Error while creating the first message from party ${this.partyIdx}: ${e}`;
+    }
+  }
+
+  getSignature(): DklsSignature {
+    if (!this.signature) {
+      throw Error('Can not request signature. Signature not produced yet.');
+    }
+    return this.signature;
+  }
+
+  endSession(): void {
+    if (this.signature) {
+      new Error('Session already ended because combined signature was produced.');
+    }
+    if (this.dsgSession) {
+      this.dsgSession.free();
+    }
+  }
+
+  handleIncomingMessages(messagesForIthRound: DeserializedMessages): DeserializedMessages {
+    let nextRoundMessages: Message[] = [];
+    let nextRoundDeserializedMessages: DeserializedMessages = { broadcastMessages: [], p2pMessages: [] };
+    this._restoreSession();
+    if (!this.dsgSession) {
+      throw Error('Session not initialized');
+    }
+    try {
+      if (this.dsgState === DsgState.Round4) {
+        const combineResult = this.dsgSession.combine(
+          messagesForIthRound.broadcastMessages.map((m) => new Message(m.payload, m.from, undefined))
+        );
+        this.signature = {
+          R: combineResult[0],
+          S: combineResult[1],
+        };
+        this.dsgState = DsgState.Complete;
+        return { broadcastMessages: [], p2pMessages: [] };
+      } else {
+        nextRoundMessages = this.dsgSession.handleMessages(
+          messagesForIthRound.broadcastMessages
+            .map((m) => new Message(m.payload, m.from, undefined))
+            .concat(messagesForIthRound.p2pMessages.map((m) => new Message(m.payload, m.from, m.to)))
+        );
+      }
+      if (this.dsgState === DsgState.Round3) {
+        nextRoundMessages = [this.dsgSession.lastMessage(Buffer.from(this.messageHash, 'hex'))];
```

Comment:
I'd prefer `messageHash` to be raw binary, and have the client of this class handle decoding from other representations.

Reply:
done